streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -412,29 +410,26 @@ protected void flushState() {
      */
     private void commitOffsets(final boolean startNewTransaction) {
         try {
-            if (commitOffsetNeeded) {
-                log.trace("Committing offsets");
-                final Map<TopicPartition, OffsetAndMetadata> consumedOffsetsAndMetadata = new HashMap<>(consumedOffsets.size());
-                for (final Map.Entry<TopicPartition, Long> entry : consumedOffsets.entrySet()) {
-                    final TopicPartition partition = entry.getKey();
-                    final long offset = entry.getValue() + 1;
-                    consumedOffsetsAndMetadata.put(partition, new OffsetAndMetadata(offset));
-                    stateMgr.putOffsetLimit(partition, offset);
-                }
-
-                if (eosEnabled) {
-                    producer.sendOffsetsToTransaction(consumedOffsetsAndMetadata, applicationId);
-                    producer.commitTransaction();
-                    transactionInFlight = false;
-                    if (startNewTransaction) {
-                        producer.beginTransaction();
-                        transactionInFlight = true;
-                    }
-                } else {
-                    consumer.commitSync(consumedOffsetsAndMetadata);
+            log.trace("Committing offsets");
+            final Map<TopicPartition, OffsetAndMetadata> consumedOffsetsAndMetadata = new HashMap<>(consumedOffsets.size());
+            for (final Map.Entry<TopicPartition, Long> entry : consumedOffsets.entrySet()) {
+                final TopicPartition partition = entry.getKey();
+                final long offset = entry.getValue() + 1;
+                consumedOffsetsAndMetadata.put(partition, new OffsetAndMetadata(offset));
+                stateMgr.putOffsetLimit(partition, offset);
+            }
+            if (eosEnabled && transactionInFlight) {
```

Comment:
Do we need `transactionInFlight` here?  Previously we just considered if `eosEnabled` for committing a transaction.

Reply:
Wondering the same. Is EOS is enabled, I would assume that `transactionInFlight` is always `true` when we hit this point?

Reply:
@bbejeck @mjsax . This fixed is aimed at test case `shouldCloseProducerOnCloseWhenEosEnabled`.
When we are trying to close `statelesstask` is with parameters `clean:true` and `isZonble:false`, 
1) It goes to `StreamTask#commitOffsets` and `transactionInFlight ` flag is not present in condition check , it is calling `producer.sendOffsetsToTransaction(consumedOffsetsAndMetadata, applicationId);`
2) And call goes to `MockProducer#sendOffsetsToTransaction` . And it verifies for `      verifyNoTransactionInFlight();`. As `transactionInFlight =false`, it throwing `There is no open transaction.` Exception.
3) So included `transactionInFlight ==true` check to filter stateless tasks with clean close.
4) Earlier this piece was not not called due to `commitOffsetNeeded ==false` for this case.

Reply:
For `shouldCloseProducerOnCloseWhenEosEnabled`, we could update it by adding `task.initializeTopology()` before `task.close()`. The former would call `producer.beginTransaction` which will then fix this issue.

Reply:
Fixed